app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java

```
@@ -0,0 +1,573 @@
+
+        writeStreamToFile(bos, newFilePath, fileEncryptionKey);
+        HybridFileBackedSqlHelpers.unsetFileAsOrphan(db, newFilePath);
+    }
+
+    @Override
+    public void remove(int id) {
+        SQLiteDatabase db = getDbOrThrow();
+
+        String filename = HybridFileBackedSqlHelpers.getEntryFilename(helper, table, id);
+        db.beginTransaction();
+        try {
+            db.delete(table, DatabaseHelper.ID_COL + "=?", new String[]{String.valueOf(id)});
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+
+        if (filename != null) {
+            File dataFile = new File(filename);
+            dataFile.delete();
+        }
+    }
+
+    @Override
+    public void remove(List<Integer> ids) {
+        if (ids.size() > 0) {
+            SQLiteDatabase db = getDbOrThrow();
+            db.beginTransaction();
+            List<String> filesToRemove;
+            try {
+                filesToRemove = HybridFileBackedSqlHelpers.getFilesToRemove(ids, helper, table);
+                List<Pair<String, String[]>> whereParamList = AndroidTableBuilder.sqlList(ids);
+                for (Pair<String, String[]> whereParams : whereParamList) {
+                    String whereClause = DatabaseHelper.ID_COL + " IN " + whereParams.first;
+                    db.delete(table, whereClause, whereParams.second);
+                }
+                db.setTransactionSuccessful();
+            } finally {
+                db.endTransaction();
+            }
+
+            HybridFileBackedSqlHelpers.removeFiles(filesToRemove);
+        }
+    }
+
+    @Override
+    public void removeAll() {
+        SQLiteDatabase db = getDbOrThrow();
+
+        db.beginTransaction();
+        try {
+            db.delete(table, null, null);
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+
+        FileUtil.deleteFileOrDir(dbDir);
```

It's better for the db entry to persist and to try and silently fail to delete non existent files than for the db entries to disappear and the files to remain, though. We can keep trying to delete blank files, but if we remove the index of the files we need to wipe, we won't even know that we need to wipe anything.